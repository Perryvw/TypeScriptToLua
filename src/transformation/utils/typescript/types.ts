import * as ts from "typescript";
import { TransformationContext } from "../../context";

export function isTypeWithFlags(context: TransformationContext, type: ts.Type, flags: ts.TypeFlags): boolean {
    if (type.symbol) {
        const baseConstraint = context.checker.getBaseConstraintOfType(type);
        if (baseConstraint && baseConstraint !== type) {
            return isTypeWithFlags(context, baseConstraint, flags);
        }
    }

    if (type.isUnion()) {
        return type.types.every(t => isTypeWithFlags(context, t, flags));
    }

    if (type.isIntersection()) {
        return type.types.some(t => isTypeWithFlags(context, t, flags));
    }

    return (type.flags & flags) !== 0;
}

export function isStringType(context: TransformationContext, type: ts.Type): boolean {
    return isTypeWithFlags(context, type, ts.TypeFlags.String | ts.TypeFlags.StringLike | ts.TypeFlags.StringLiteral);
}

export function isNumberType(context: TransformationContext, type: ts.Type): boolean {
    return isTypeWithFlags(context, type, ts.TypeFlags.Number | ts.TypeFlags.NumberLike | ts.TypeFlags.NumberLiteral);
}

function isExplicitArrayType(context: TransformationContext, type: ts.Type): boolean {
    if (type.symbol) {
        const baseConstraint = context.checker.getBaseConstraintOfType(type);
        if (baseConstraint && baseConstraint !== type) {
            return isExplicitArrayType(context, baseConstraint);
        }
    }

    if (type.isUnionOrIntersection()) {
        return type.types.some(t => isExplicitArrayType(context, t));
    }

    const flags = ts.NodeBuilderFlags.InTypeAlias | ts.NodeBuilderFlags.AllowEmptyTuple;
    let typeNode = context.checker.typeToTypeNode(type, undefined, flags);
    if (typeNode && ts.isTypeOperatorNode(typeNode) && typeNode.operator === ts.SyntaxKind.ReadonlyKeyword) {
        typeNode = typeNode.type;
    }

    return typeNode !== undefined && (ts.isArrayTypeNode(typeNode) || ts.isTupleTypeNode(typeNode));
}

/**
 * Iterate over a type and its bases until the callback returns true.
 */
function forTypeOrAnySupertype(
    context: TransformationContext,
    type: ts.Type,
    predicate: (type: ts.Type) => boolean
): boolean {
    if (predicate(type)) {
        return true;
    }

    if (!type.isClassOrInterface() && type.symbol) {
        type = context.checker.getDeclaredTypeOfSymbol(type.symbol);
    }

    return (type.getBaseTypes() ?? []).some(superType => forTypeOrAnySupertype(context, superType, predicate));
}

export function isArrayType(context: TransformationContext, type: ts.Type): boolean {
    return forTypeOrAnySupertype(context, type, t => isExplicitArrayType(context, t));
}

export function isFunctionType(context: TransformationContext, type: ts.Type): boolean {
    const typeNode = context.checker.typeToTypeNode(type, undefined, ts.NodeBuilderFlags.InTypeAlias);
    return typeNode !== undefined && ts.isFunctionTypeNode(typeNode);
}
