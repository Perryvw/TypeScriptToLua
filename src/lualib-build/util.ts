import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
