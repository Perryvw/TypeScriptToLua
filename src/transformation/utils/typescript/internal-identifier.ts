import * as ts from "typescript";
import * as lua from "../../../LuaAST";

export interface InternalIdentifierData {
    isTransformedOptionalBase?: boolean;
    optionalBaseContextualCall?: lua.CallExpression;
}

const internalIdentifiers = new WeakMap<ts.Identifier, InternalIdentifierData>();

// an internal identifier will be transformed verbatim to lua
export function createInternalIdentifier(
    text: string,
    tsOriginal?: ts.Expression,
    data: InternalIdentifierData = {}
): ts.Identifier {
    const identifier = ts.factory.createIdentifier(text);
    ts.setOriginalNode(identifier, tsOriginal);
    internalIdentifiers.set(identifier, data);
    return identifier;
}
export function isInternalIdentifier(identifier: ts.Identifier): boolean {
    return internalIdentifiers.has(identifier);
}

export function getInternalIdentifierData(identifier: ts.Identifier): InternalIdentifierData | undefined {
    return internalIdentifiers.get(identifier);
}
