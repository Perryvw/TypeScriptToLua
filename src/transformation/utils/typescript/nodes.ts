import * as ts from "typescript";
import { TransformationContext } from "../../context";

export function isAssignmentPattern(node: ts.Node): node is ts.AssignmentPattern {
    return ts.isObjectLiteralExpression(node) || ts.isArrayLiteralExpression(node);
}

export function isDestructuringAssignment(node: ts.Node): node is ts.DestructuringAssignment {
    return (
        ts.isBinaryExpression(node) &&
        node.operatorToken.kind === ts.SyntaxKind.EqualsToken &&
        isAssignmentPattern(node.left)
    );
}

export function isAmbientNode(node: ts.Declaration): boolean {
    return (ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Ambient) !== 0;
}

export function isInDestructingAssignment(node: ts.Node): boolean {
    return (
        node.parent &&
        ((ts.isVariableDeclaration(node.parent) && ts.isArrayBindingPattern(node.parent.name)) ||
            (ts.isBinaryExpression(node.parent) && ts.isArrayLiteralExpression(node.parent.left)))
    );
}

export function getNodeSymbol(context: TransformationContext, node: ts.Node): ts.Symbol | undefined {
    return (node as any).symbol ?? context.checker.getSymbolAtLocation(node);
}

export function isFirstDeclaration(context: TransformationContext, node: ts.Node) {
    const symbol = getNodeSymbol(context, node);
    return symbol ? symbol.valueDeclaration === node : true;
}
