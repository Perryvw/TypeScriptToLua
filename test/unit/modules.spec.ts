import * as ts from "typescript";
import * as util from "../util";

describe("export default", () => {
    test("literal", () => {
        util.testModule`
            export default true;
        `.expectToEqual({ default: true });
    });

    test("class", () => {
        util.testModule`
            export default class Default {}
            const d = new Default();
            export const result = d.constructor.name;
        `
            .setReturnExport("result")
            .expectToMatchJsResult();
    });

    test("function", () => {
        util.testModule`
            export default function defaultFunction() {
                return true;
            }
            export const result = defaultFunction();
        `
            .setReturnExport("result")
            .expectToMatchJsResult();
    });
});

test("export { value as default }", () => {
    util.testBundle`
        const value = true;
        export { value as default };
    `.expectToEqual({ default: true });
});

test("export { default } from '...'", () => {
    util.testBundle`
        export { default } from "./module";
    `
        .addExtraFile("module.ts", "export default true;")
        .expectToEqual({ default: true });
});

test("import = and export =", () => {
    util.testBundle`
        import m = require("./module");
        export const value = m;
    `
        .setOptions({ module: ts.ModuleKind.CommonJS })
        .addExtraFile("module.ts", "export = true;")
        .expectToEqual({ value: true });
});

test("Mixed Exports, Default and Named Imports", () => {
    util.testBundle`
        import defaultExport, { a, b, c } from "./module";
        export const value = defaultExport + b + c;
    `
        .addExtraFile(
            "module.ts",
            `
                export const a = 1;
                export const b = 2;
                export const c = 3;
                export default a;
            `
        )
        .expectToEqual({ value: 6 });
});

test("Mixed Exports, Default and Namespace Import", () => {
    util.testBundle`
        import defaultExport, * as ns from "./module";
        export const value = defaultExport + ns.b + ns.c;
    `
        .addExtraFile(
            "module.ts",
            `
                export const a = 1;
                export const b = 2;
                export const c = 3;
                export default a;
            `
        )
        .expectToEqual({ value: 6 });
});

describe("export live bindings", () => {
    const reassignmentTestCases = [
        "x = 1",
        "x++",
        "(x = 1)",
        "[x] = [1]",
        "[[x]] = [[1]]",
        "({ x } = { x: 1 })",
        "({ y: x } = { y: 1 })",
        "({ x = 1 } = { x: undefined })",
    ];

    // https://github.com/TypeScriptToLua/TypeScriptToLua/issues/926
    test.each(reassignmentTestCases.filter(c => !c.includes(" } = { x: ")))("export variable (%p)", reassignment => {
        util.testModule`
            export let x = 0;
            ${reassignment};
        `.expectToMatchJsResult();
    });

    test.each(reassignmentTestCases)("export variable as a binding (%p)", reassignment => {
        util.testModule`
            let x = 0;
            export { x };
            ${reassignment};
        `.expectToMatchJsResult();
    });

    test.each(reassignmentTestCases)("export variable with multiple bindings (%p)", reassignment => {
        util.testModule`
            let x = 0;
            export { x as a };
            export { x as b };
            ${reassignment};
        `.expectToMatchJsResult();
    });

    // Can't be added to reassignmentTestCases because of https://github.com/microsoft/TypeScript/issues/35881
    test("export variable (for in loop)", () => {
        util.testModule`
            export let foo = '';
            for (foo in { x: true }) {}
        `
            .setReturnExport("foo")
            .expectToMatchJsResult();
    });

    test("export variable as a binding (for in loop)", () => {
        util.testModule`
            let foo = '';
            export { foo as bar };
            for (foo in { x: true }) {}
        `
            .setReturnExport("bar")
            .expectToEqual("x");
    });

    test("does not update shadowed names", () => {
        util.testModule`
            export let a = 1;
            { let a = 2; a = 3 };
        `.expectToMatchJsResult();
    });

    test("renamed export specifier shouldn't effect local vars", () => {
        util.testModule`
            let x = false;
            export { x as a };
            let a = 5;
            a = 6;
        `.expectToMatchJsResult();
    });
});

describe("import and export elision", () => {
    const moduleDeclaration = `
        declare module "module" {
            export type Type = string;
            export declare const value: string;
        }
    `;

    const expectToElideImport: util.TapCallback = builder => {
        builder.addExtraFile("module.d.ts", moduleDeclaration).setOptions({ module: ts.ModuleKind.CommonJS });
        expect(builder.getLuaExecutionResult()).not.toBeInstanceOf(util.ExecutionError);
    };

    test("should elide named type imports", () => {
        util.testModule`
            import { Type } from "module";
            const foo: Type = "bar";
        `.tap(expectToElideImport);
    });

    test("should elide named value imports used only as a type", () => {
        util.testModule`
            import { value } from "module";
            const foo: typeof value = "bar";
        `.tap(expectToElideImport);
    });

    test("should elide namespace imports with unused values", () => {
        util.testModule`
            import * as module from "module";
            const foo: module.Type = "bar";
        `.tap(expectToElideImport);
    });

    test("should elide `import =` declarations", () => {
        util.testModule`
            import module = require("module");
            const foo: module.Type = "bar";
        `.tap(expectToElideImport);
    });

    test("should elide type exports", () => {
        util.testModule`
            (globalThis as any).foo = true;
            type foo = boolean;
            export { foo };
        `.expectToEqual([]);
    });
});

test.each(["ke-bab", "dollar$", "singlequote'", "hash#", "s p a c e", "ɥɣɎɌͼƛಠ", "_̀ः٠‿", ".dot"])(
    "import local identifier generation (%p)",
    name => {
        util.testBundle`
            import { foo } from "./${name}";
            export { foo };
        `
            .addExtraFile(`${name}.ts`, "export const foo = true;")
            .expectToEqual({ foo: true })
            .tap(builder => {
                const identifier = builder.getMainLuaCodeChunk().match(/local (.+) = require\(/)?.[1];
                expect(identifier).toMatchSnapshot("local identifier");
            });
    }
);
