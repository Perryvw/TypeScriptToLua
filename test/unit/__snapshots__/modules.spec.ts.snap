// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import local identifier generation (".dot"): local identifier 1`] = `"_____2Edot"`;

exports[`import local identifier generation ("_̀ः٠‿"): local identifier 1`] = `"______300_903_660_203F"`;

exports[`import local identifier generation ("dollar$"): local identifier 1`] = `"____dollar_24"`;

exports[`import local identifier generation ("hash#"): local identifier 1`] = `"____hash_23"`;

exports[`import local identifier generation ("ke-bab"): local identifier 1`] = `"____ke_2Dbab"`;

exports[`import local identifier generation ("s p a c e"): local identifier 1`] = `"____s_20p_20a_20c_20e"`;

exports[`import local identifier generation ("singlequote'"): local identifier 1`] = `"____singlequote_27"`;

exports[`import local identifier generation ("ɥɣɎɌͼƛಠ"): local identifier 1`] = `"_____265_263_24E_24C_37C_19B_CA0"`;
